### PR TITLE
Update fetcher.ts to work like the browser

### PR DESCRIPTION
**Pull request by menduz**
This PR  removes a forceful throw and `return undefined` to copy the default behavior of the browser.

---

### Context

I'm trying to use 
```ts
if(request.status == 400) { 
  ...
}
```
but the 
```
if !request.ok then throw
```
was preventing me from doing it

### CODE REVIEW

**Comment by menduz** [comment]
@juanmahidalgo @aleortega if you are OK with the approach, I'll correct the tests. 

With retries, I'd return the last try's response

File: src/fetcher.ts [hunk 1]
```diff
@@ -80,12 +80,6 @@ export function createFetchComponent(defaultOptions?: FetcherOptions): IFetchCom
       abortController: controller
     })
 
-    // Throw in case of error
-    if (!response?.ok) {
-      const responseText = await response?.text()
-      throw new Error(`Failed to fetch ${url}. Got status ${response?.status}. Response was '${responseText}'`)
-    }
-
     if (!isUsingNode() && !!response) {
```
Comment (juanmahidalgo) [inline src/fetcher.ts:83]:
```suggestion
    if (!isUsingNode() && !!response && response?.ok) {
```
do we need to add `response?.ok` here so it creates the buffer just if `response.ok` is truthy?

Comment (juanmahidalgo) [reply src/fetcher.ts:83]: yeah, true, they should return the buffer as well. LGTM then!!

**Comment by aleortega** [comment]
> @juanmahidalgo @aleortega if you are OK with the approach, I'll correct the tests.
> 
> With retries, I'd return the last try's response

Hello, looks good to me as well, in fact this was my first approach back then but received a feedback from the team saying that they preferred to throw in case of `StatusCode !== 2xx`. Not sure if it is okay adding an option to force throws or in this case (_to avoid breaking changes_), prevent throwing if a specific option is provided. What do you think?

Anyways, this is not a strong opinion but may this change will bring us the versatility to opt-in/out the preferred approach.

**Comment by menduz** [comment]
I think it is a ..weird behavior, I was trying to get the error message and fields from the 401 response; and instead I got a hard to parse exception message. 

In the long term, I think the best is not to assume that this "throw" is a good thing and do server code as we do in the front-end, checking for the response.ok every time depending on the use case. That way anyone can leverage their knowledge about how "fetch" works, without any subtlety bound to this implementation. Otherwise, I'd advice to name the function differently to "fetch" following the maxima of functional programming "if the function does, receives or returns something different, it should be named differently".

**Comment by aleortega** [comment]
Got it, so I'm ok with merging this. I'll update the code using this fetch to adapt it to the new implementation.

Thanks for the heads-up!

**Review by aleortega — APPROVED** [review]

**Comment by hugoArregui** [comment]
@menduz are you going to work on the tests? I want to integrate this module but I would like to have this merged first, thanks!

**Comment by aleortega** [comment]
> [...] I had already adapted these tests in the `feat/optin-prevent-throwing` branch [...]

In case you're curious about the other approach, here is [the PR](https://github.com/well-known-components/fetch-component/pull/67).

**Comment by menduz** [comment]
thanks for the context, the tests and the experiment Ale! I also think it looks cleaner without the opt-in.

I think we are go for mering and releasing a new major. le damos?

**Comment by aleortega** [comment]
> I think we are go for mering and releasing a new major. le damos?

Yass, all yours :)  I'll let the team know about the change and start upgrading some services where I can use this (eager to remove those try/catches).
